src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,448 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using static System.Runtime.InteropServices.JavaScript.JSHostImplementation;
+
+namespace System.Runtime.InteropServices.JavaScript
+{
+    internal sealed class JSProxyContext : IDisposable
+    {
+        private bool _disposedValue;
+
+        // we use this to maintain identity of JSHandle for a JSObject proxy
+        private readonly Dictionary<nint, WeakReference<JSObject>> ThreadCsOwnedObjects = new();
+        // we use this to maintain identity of GCHandle for a managed object
+        private readonly Dictionary<object, nint> ThreadJsOwnedObjects = new(ReferenceEqualityComparer.Instance);
+        // this is similar to GCHandle, but the GCVHandle is allocated on JS side and this keeps the C# proxy alive
+        private readonly Dictionary<nint, PromiseHolder> ThreadJsOwnedHolders = new();
+        // JSVHandle is like JSHandle, but it's not tracked and allocated by the JS side
+        // It's used when we need to create JSHandle-like identity ahead of time, before calling JS.
+        // they have negative values, so that they don't collide with JSHandles.
+        private nint NextJSVHandle = -2;
+        private readonly List<nint> JSVHandleFreeList = new();
+
+#if !FEATURE_WASM_THREADS
+        public static readonly JSProxyContext DefaultInstance = new();
+        public static JSProxyContext CurrentInstance => DefaultInstance;
+        public static JSProxyContext CapturedInstance
+        {
+            get => MainInstance!;
+            set { }
+        }
+        public static JSProxyContext MainInstance => DefaultInstance;
+#else
+        public nint TargetTID;
+        public int ThreadId;
+        public bool IsMainThread;
+        public JSSynchronizationContext SynchronizationContext;
+
+        [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "thread_id")]
+        private static extern ref long GetThreadNativeThreadId(Thread @this);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IntPtr GetNativeThreadId()
+        {
+            return (int)GetThreadNativeThreadId(Thread.CurrentThread);
```

Comment:
is this meant to return IntPtr, int, or nint? There are types being mixed here

Reply:
the underlying managed field is `long` and the wasm32 value is always 32bits.
for sake of convenience I use IntPtr/nint data type to hold it in all managed places.

this is being passed as parameter to icalls, that's why i preferred 32bits. Should I switch it to long ? Does that have zero extra cost ?

Reply:
discussed elsewhere, will keep it IntPtr, because that matches `pthread_t` and `MonoNativeThreadId` parameter of C function I need to call.